### PR TITLE
[REG-762] - Forbid domain names minting with leading and ending hyphens

### PR DESCRIPTION
**Pull request by DeRain**
### 1. Contracts versioning
- [x] Make sure that the `patch` version of the contracts is increased if changes have been made to the `UNSRegistry`, `MintingManager` or `ProxyReader` contracts.
- [x] Make sure that the `minor` version of the contracts is increased if breaking changes have been made to the `UNSRegistry`, `MintingManager` or `ProxyReader` contracts. It includes changes of interfaces.
### 2. Contracts licensing
- [x] Make sure that no **SPDX-License-Identifier** defined in contracts.
- [x] Make sure that the **header** is added to the new contract files. 
  ```
  // @author Unstoppable Domains, Inc.
  // @date {Month} {Day}(ordinal), {Year}
  ```
### 3. Coverage
- [x] Make sure that the coverage of contracts has not decreased and strive **100%**
### 4. Configs versioning
- [x] Make sure that the version of `uns-config.json` is increased if changes have been made to the config.
- [x] Make sure that the version of `resolver-keys.json` is increased if changes have been made to the config.
### 5. Package versioning
- [x] Make sure that the `patch` version of package is increased if valuable changes have been made to the package. It includes contracts update, configs update, etc.
- [x] Make sure that the `major.minor` version of package is synced with version of `UNSRegistry` contract.
- [x] Make sure that the `CHANGELOG` is updated with short description for the new version. 
### 6. Code review
- [ ] `resolver-keys.json` code review is required from **DevTools** team
- [ ] For all other changes code review is required from **Registry** team

### CODE REVIEW

File: contracts/MintingManager.sol [hunk 1]
```diff
@@ -323,10 +323,15 @@ contract MintingManager is ERC2771Context, MinterRole, Blocklist, Pausable, IMin
             ptr := add(str, 0x20)
         }
 
-        for (uint256 i = 0; i < bytes(str).length; i++) {
+        uint8 hyphen = 45;
+        uint256 labelLength = bytes(str).length;
+        if (_charAt(ptr, 0) == hyphen || _charAt(ptr, labelLength - 1) == hyphen) {
+            return false;
+        }
+        for (uint256 i = 0; i < labelLength; i++) {
```
Comment (aquiladev) [inline contracts/MintingManager.sol:331]: it does not make any sense to check first and last char twice, take a look the [PR](https://github.com/unstoppabledomains/uns/pull/202), there are some commits which cover this logic, but without double check

File: test/MintingManager.test.ts [hunk 2]
```diff
@@ -469,7 +469,22 @@ describe('MintingManager', () => {
       const issueWithRecordsWithReverse = 'issueWithRecords(address,string[],string[],string[],bool)';
 
       it('should not allow to mint domains with invalid labels', async () => {
-        const labels = ['', ' ', 'Q', 'qwertY', 'qwErty', '*', '$3123', '\tq', 'q\n', '@', 'k.h'];
+        const labels = [
```
Comment (aquiladev) [inline test/MintingManager.test.ts:472]: I would add invalid cases '-', '--' as well

File: contracts/MintingManager.sol [hunk 3]
```diff
@@ -323,10 +323,15 @@ contract MintingManager is ERC2771Context, MinterRole, Blocklist, Pausable, IMin
             ptr := add(str, 0x20)
         }
 
-        for (uint256 i = 0; i < bytes(str).length; i++) {
+        uint8 hyphen = 45;
+        uint256 labelLength = bytes(str).length;
+        if (_charAt(ptr, 0) == hyphen || _charAt(ptr, labelLength - 1) == hyphen) {
```
Comment (aquiladev) [inline contracts/MintingManager.sol:328]: what it there is one-symbol label?

Comment (DeRain) [reply contracts/MintingManager.sol:328]: In this case two conditions will check the same character

File: contracts/MintingManager.sol [hunk 4]
```diff
@@ -324,7 +324,22 @@ contract MintingManager is ERC2771Context, MinterRole, Blocklist, Pausable, IMin
...
+            if (
+                !(charData >= 48 && charData <= 57) && !(charData >= 97 && charData <= 122) // 0-9 and a-z
+            ) {
+                return false;
+            }
+        }
+        for (uint256 i = 1; i < labelLength - 1; i++) {
             uint8 data = _charAt(ptr, i);
             if (
                 data != 45 && // hyphen (-)
```
Comment (bogdan) [inline contracts/MintingManager.sol:345]: Checking first and last symbol using `i == 0 || i == length - 1` can be more efficient in terms of contract size.

**Review by bogdan — APPROVED** [review]

File: contracts/MintingManager.sol (hunk 4 above)
Comment (DeRain) [reply contracts/MintingManager.sol:345]: @bogdan I was thinking about this implementation but then decided to check first and last symbols only once to save some gas on runtime.

**Review by aquiladev — APPROVED** [review]

**Review by bogdan — APPROVED** [review]

**Pull request merged**